esp-hal/src/dma/mod.rs

```diff
@@ -366,6 +366,26 @@ impl DmaDescriptor {
             true => Owner::Dma,
         }
     }
+
+    fn iter(&self) -> impl Iterator<Item = &DmaDescriptor> {
+        core::iter::successors(Some(self), |d| {
+            if d.next.is_null() {
+                None
+            } else {
+                Some(unsafe { &*d.next })
+            }
+        })
```

Comment:
You can simplify this to
```suggestion
        core::iter::successors(Some(self), |d| unsafe { d.next.as_ref() })
```

You can use `as_mut` for the lower one.